// terriajs/dist/Core/Json.js
export function isJsonObject(value) {
    return (value !== undefined &&
        value !== null &&
        typeof value === "object" &&
        !Array.isArray(value));
}
export function isJsonBoolean(value) {
    return typeof value === "boolean";
}
export function isJsonNumber(value) {
    return typeof value === "number";
}
export function isJsonString(value) {
    return typeof value === "string";
}
export function assertObject(value, name) {
    if (isJsonObject(value))
        return;
    throwUnexpectedError("JsonObject", typeof value, name);
}
export function assertString(value, name) {
    if (typeof value === "string")
        return;
    throwUnexpectedError("string", typeof value, name);
}
export function assertNumber(value, name) {
    if (typeof value === "number")
        return;
    throwUnexpectedError("number", typeof value, name);
}
export function assertArray(value, name) {
    if (Array.isArray(value))
        return;
    throwUnexpectedError("Array", typeof value, name);
}
function throwUnexpectedError(expectedType, actualType, name) {
    const nameToBe = name ? ` ${name} to be` : "";
    throw new Error(`Expected${nameToBe} ${expectedType}, got ${actualType}`);
}
